types/basetypes/dynamic_value.go

```diff
@@ -0,0 +1,151 @@
+
+	// DynamicSemanticEquals should return true if the given value is
+	// semantically equal to the current value. This logic is used to prevent
+	// Terraform data consistency errors and resource drift where a value change
+	// may have inconsequential differences.
+	//
+	// Only known values are compared with this method as changing a value's
+	// state implicitly represents a different value.
+	DynamicSemanticEquals(context.Context, DynamicValuable) (bool, diag.Diagnostics)
+}
+
+// NewDynamicValue creates a Dynamic with a known value. Access the value via the Dynamic
+// type UnderlyingValue method. The concrete value type returned to Terraform from this value
+// will be determined by the provided `(attr.Value).ToTerraformValue` function.
+func NewDynamicValue(value attr.Value) DynamicValue {
+	return DynamicValue{
+		value: value,
+		state: attr.ValueStateKnown,
+	}
+}
+
+// NewDynamicNull creates a Dynamic with a null value. The concrete value type returned to Terraform
+// from this value will be tftypes.DynamicPseudoType.
+func NewDynamicNull() DynamicValue {
+	return DynamicValue{
+		state: attr.ValueStateNull,
+	}
+}
+
+// NewDynamicUnknown creates a Dynamic with an unknown value. The concrete value type returned to Terraform
+// from this value will be tftypes.DynamicPseudoType.
+func NewDynamicUnknown() DynamicValue {
+	return DynamicValue{
+		state: attr.ValueStateUnknown,
+	}
+}
+
+// DynamicValue represents a dynamic value. Static types are always
+// preferable over dynamic types in Terraform as practitioners will receive less
+// helpful configuration assistance from validation error diagnostics and editor
+// integrations.
+type DynamicValue struct {
+	// value contains the known value, if not null or unknown.
+	value attr.Value
+
+	// state represents whether the value is null, unknown, or known. The
+	// zero-value is null.
+	state attr.ValueState
+}
+
+// Type returns DynamicType.
+func (v DynamicValue) Type(ctx context.Context) attr.Type {
+	return DynamicType{}
+}
+
+// ToTerraformValue returns the equivalent tftypes.Value for the DynamicValue.
+func (v DynamicValue) ToTerraformValue(ctx context.Context) (tftypes.Value, error) {
+	switch v.state {
+	case attr.ValueStateKnown:
+		return v.value.ToTerraformValue(ctx)
+	case attr.ValueStateNull:
+		return tftypes.NewValue(tftypes.DynamicPseudoType, nil), nil
+	case attr.ValueStateUnknown:
+		return tftypes.NewValue(tftypes.DynamicPseudoType, tftypes.UnknownValue), nil
+	default:
+		panic(fmt.Sprintf("unhandled Dynamic state in ToTerraformValue: %s", v.state))
+	}
+}
+
+// Equal returns true if the given attr.Value is also a DynamicValue and contains an equal underlying value as defined by its Equal method.
+func (v DynamicValue) Equal(o attr.Value) bool {
+	other, ok := o.(DynamicValue)
+	if !ok {
+		return false
+	}
+
+	if v.state != other.state {
+		return false
+	}
+
+	if v.state != attr.ValueStateKnown {
+		return true
+	}
+
+	return v.value.Equal(other.value)
```

Comment:
Should I double check this isn't `nil`? Return `false`?

Reply:
Yeah this sounds like it might need some awkward additional `nil` checks just to prevent potential panics and associated bug reports. I would treat this case similar to the upstream `(tftypes.Value).Equal()` panic removal where folks should be able to inspect the values more if they are getting "unexpected" inequality.

Reply:
Forced inequality if either of the underlying values are `nil`. Since it's an invalid state, I decided to determine they are always not equal (even if both underlying values are `nil`)

6bb2360 ⚡